### ISSUE TITLE
[7.0.0rc2 regression] `JavaInfo` can no longer be stored in `depset`

### ISSUE THREAD

**Issue by fmeum** [bazelbuild/bazel#19999]
### Description of the bug:

Since https://github.com/bazelbuild/bazel/commit/9320a217779208aece2c8995ceb13d18be66f728, `JavaInfo` instances are no longer Starlark immutable and thus can't be stored in `depset`s, which is a breaking change (e.g. [this aspect](https://github.com/bazeltools/bazel_jar_jar/blob/main/jar_jar_aspect.bzl) is broken by it).

### What's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.

```starlark
# WORKSPACE

# BUILD
load(":rule.bzl", "my_rule")

my_rule(name = "my_rule")

# rule.bzl
def my_rule_impl(ctx):
    file = ctx.actions.declare_file("file.jar") 
    ctx.actions.write(file, "foo")
    depset([JavaInfo(compile_jar = file, output_jar = file)])

my_rule = rule(my_rule_impl)
```

Then `bazel build //...` fails with 9320a217779208aece2c8995ceb13d18be66f728 and later, while passing before the commit.

### Which operating system are you running Bazel on?

Any

**Comment by fmeum** [comment]
CC @hvadehra @comius I suspect that this may also affect other Starlarkified providers.

**Comment by fmeum** [comment]
@bazel-io flag

**Comment by hvadehra** [comment]
Thanks for reporting. Fix incoming shortly.

FWIW, ordinarily this should not be an issue since Starlark providers returned from a target/aspect are frozen (and hence immutable) once analysis of the target is complete. In this case, JavaInfo Starlarkification needed native/starlark provider interop, and I missed explicitly marking some native providers (that are also `StarlarkValue`s) as immutable)

**Comment by hvadehra** [comment]
For completeness, the above explanation does not apply to your repro example (i.e. providers created _in situ_), but I don't expect a genuine case to have to do this. But the somewhat long-term plan is to make JavaInfo (and other providers) deeply immutable by design. This will mainly involve changing all list types to depsets, but that is a breaking change in itself, and I expect the migration to be long and arduous.

**Comment by fmeum** [comment]
> This will mainly involve changing all list-type fields to depsets. That is a breaking change in itself, and I expect the migration to be long and arduous.

Thanks for providing the full context. Would it be an option to migrate to tuples first, potentially resulting in a smaller migration?

**Comment by hvadehra** [comment]
Yes, I imagine it would certainly help. 

At least one case where that doesn't work is if a (currently list) value is concatenated with some other list, leading to `Error: unsupported binary operation: tuple + list`

**Comment by iancha1992** [comment]
@bazel-io fork 7.0.0

**Comment by iancha1992** [comment]
A fix for this issue has been included in [Bazel 7.0.0 RC5](https://releases.bazel.build/7.0.0/rc5/index.html). Please test out the release candidate and report any issues as soon as possible. Thanks!
